### ISSUE TITLE
CRDGenerator v2: Gradle Plugin

### ISSUE THREAD

**Issue by baloo42** [fabric8io/kubernetes-client#5945]
## Description

CRD Generator v1 and its annotation processor (`crd-generator-apt`) were deprecated in
7.0.0 and removed in 8.0.0 (#8031). Maven users have `crd-generator-maven-plugin`, but
Gradle users have no first-class option — only the build-script recipes in
[crd-generator/gradle/README.md](https://github.com/fabric8io/kubernetes-client/blob/main/crd-generator/gradle/README.md)
(generate in the Gradle daemon, or fork a `JavaExec` with `crd-generator-cli` when the
compile JDK differs from the daemon JDK).

The plugin should generate CRDs from the compiled custom resource classes using the
CRD Generator v2 API, with feature parity with the Maven plugin.

## Requirements

- Java (not Kotlin), built with Maven, following `java-generator/gradle-plugin`.
- Options named after their Maven counterparts (see `CrdGeneratorMojo`): output dir,
  CRD versions (v1), parallel generation (on by default), file header, debug output.
- Use `crd-generator-collector` (Jandex) to find the custom resource classes.
- Up-to-date checks / build cache support (same gap as #7091 for java-generator).
- Task name and group consistent with the java-generator plugin.
- Docs: `doc/CRD-generator.md` Quick start still says the Gradle plugin is not available,
  and `crd-generator/gradle/README.md` needs updating once it lands.

## Status

- Draft PR #6910 (@Tanemahuta), open, needs tests.
- WIP branch: https://github.com/baloo42/kubernetes-client/tree/feature/iss_5945
- Closed by the stale bot in Sept 2025. Reopening: with v1 removed in 8.0.0, Gradle users
  now have no supported tooling at all.

**Comment by shawkins** [comment]
As suggested by @manusa we can just start with a clone of https://github.com/fabric8io/kubernetes-client/tree/main/java-generator/maven-plugin

**Comment by manusa** [comment]
As a final task, the CRD-Generator need to be updated.
https://github.com/fabric8io/kubernetes-client/pull/6674#discussion_r1862667960

**Comment by heruan** [comment]
Is the Gradle plugin already targeted for a future release?

**Comment by baloo42** [comment]
Planned to be implemented: yes. Targeted to a specific release: no.

I have started to work on it but it is not finished:
https://github.com/baloo42/kubernetes-client/tree/feature/iss_5945

I don't have any time until March, so whoever has time can jump in.

Meanwhile gradle users can use the workaround described here:

https://github.com/fabric8io/kubernetes-client/blob/main/crd-generator/gradle/README.md

**Comment by Tanemahuta** [comment]
I started the [PR#69106](https://github.com/fabric8io/kubernetes-client/pull/69106).

I needed to port my code from a gradle build (which I prefer ;) ), I also think it would be best to build gradle plugins using gradle, but that is up to the project.

Could you please have a quick look over the implementation and integration, and check, if this suites you?
I will then try to fix the remaining parts.

**Comment by Tanemahuta** [comment]
Hello there, this is a ping.
I am still willing to work on https://github.com/fabric8io/kubernetes-client/pull/6910.

**Comment by manusa** [comment]
> I also think it would be best to build gradle plugins using gradle, but that is up to the project.

It would be better and much easier to integrate with the entire Gradle ecosystem. It would also make it easier to publish (if done in isolation).

However, the gradle plugin is just a small part of the multi-module kubernetes-client project.
Making the plugin project Gradle-based would make it extremely harder to maintain and publish since everything else in the code base is Maven based.

In general, you should try to follow what was already done for https://github.com/fabric8io/kubernetes-client/tree/main/java-generator/gradle-plugin

We have a lot of tooling around Gradle to overcome the issues of not building a Gradle plugin in Gradle. So that shouldn't be a problem.

The implementation is in Kotlin which would also be a no-go since everything else is Java.

The configuration options should be named just like their [Maven counter-parts](https://github.com/fabric8io/kubernetes-client/blob/ad99f51d3fb04e4916a86b76cadcd345e5a9de9d/crd-generator/maven-plugin/src/main/java/io/fabric8/crd/generator/maven/plugin/CrdGeneratorMojo.java).

> Hello there, this is a ping. I am still willing to work on [#6910](https://github.com/fabric8io/kubernetes-client/pull/6910).

That's awesome, just let us know if you need any additional help or guidance

**Comment by heruan** [comment]
Thanks @Tanemahuta for working on this! It would be great if the two Gradle plugin would share the same task group and follow the same naming conventions. The current java-generator plugin for example has a very generic plugin name (i.e. `javaGen`) that could likely conflict with other plugins; it has then the `crd2java` task that could maybe match a `java2crd` task for CRD generation?

Or maybe think of new names, e.g. `fabric8GenerateJava` and `fabric8GenerateCRDs`?

**Comment by heruan** [comment]
And support for caching would be greatly appreciated for both tasks, so that the task wouldn't run if there are no changes to the sources.

**Comment by baloo42** [comment]
@Tanemahuta I can assist if you need some guidance, or jump in if you don't have the time anymore.

In general, I think we should finish this soon to make it a complete solution.

**Comment by HyoKImAbsolut** [comment]
any progress here?

**Comment by baloo42** [comment]
Unfortunatly not (as far as i know) - I don't need it myself nor my customers (and I'm not a gradle pro).

But i have started with an implementation a while ago.

https://github.com/baloo42/kubernetes-client/tree/feature/iss_5945

I think i got stuck wiring up integration tests. If someone wants to assist (or the other way around), I'm willing to help/finish it.

@HyoKImAbsolut Do you want to assist or take over?
@Tanemahuta Any progress on your side?

**Comment by manusa** [comment]
Reopening: CRD Generator v1 and its annotation processor were removed in 8.0.0 (#8031), so Gradle users are left with the build-script recipes in crd-generator/gradle/README.md and no first-class tooling. Refreshed the description; draft PR #6910 is still open.
